src/main/scala/com/foursquare/common/thrift/bson/ByteStringBuilder.scala

```diff
@@ -0,0 +1,115 @@
+// Copyright 2015 Foursquare Labs Inc. All Rights Reserved.
+
+package com.foursquare.common.thrift.bson
+
+import java.io.{ByteArrayInputStream, InputStream}
+import java.nio.charset.Charset
+import java.lang.reflect.Constructor
+
+private[bson] object ByteStringBuilder {
+  val UTF8_CHARSET = Charset.forName("UTF-8")
+  val ASCII_CHARSET = Charset.forName("US-ASCII")
+  val MaxSize = 16 * 1024 * 1024
+  val StringConstructor: Constructor[String] = {
+    try {
+      val constructor = {
+        classOf[String].getDeclaredConstructors.find(_.toString == "java.lang.String(char[],boolean)").get
+      }
+      constructor.setAccessible(true)
+      constructor.asInstanceOf[Constructor[String]]
+    } catch {
+      case e: Exception =>
+        System.err.println("Error reflecting for String(chars, boolean) private constructor. Will use public constructor.")
+        e.printStackTrace
+        null
+    }
+  }
+
+  val SingleByteStrings: Array[String] = (32 to 127).map(b => new String(Array(b.toChar))).toArray
+
+  def byteToString(b: Byte): String = {
+    SingleByteStrings(b - 32)
+  }
+}
+
+/**
+ * Growable buffer for building strings. Not thread safe
+ * contains a bunch of performance hacks
+ * reset() must be called before each re-use
+ */
+class ByteStringBuilder(initialSize: Int) {
+  private var bytes = new Array[Byte](initialSize)
+  private var length = 0
+  private var isAscii = true
+  
+  private def ensureGrowth(size: Int) {
+    val newRequestedSize = length + size
+    if (newRequestedSize > bytes.length) {
+      if (newRequestedSize > ByteStringBuilder.MaxSize) {
+        throw new RuntimeException(s"Attempting to grow string builder past maximum size $newRequestedSize")
+      }
+      val newLength = math.min(newRequestedSize * 1.25, ByteStringBuilder.MaxSize).toInt
+      val newBytes = new Array[Byte](newLength)
+      System.arraycopy(bytes, 0, newBytes, 0, length)
+      bytes = newBytes
+    }
+  }
+
+  private def checkAscii(b: Byte) {
+    isAscii = isAscii && b >= 32 && b <= 127
+  }
+
+  def append(b: Byte) {
+    ensureGrowth(1)
+    checkAscii(b)
+    bytes(length) = b
+    length += 1
+  }
+
+  def reset() {
+    length = 0
+    isAscii = true
+  }
+
+  def build(): String = {
+    // Let the performance hacks begin:
+    // 1. don't construct new empty string
+    // 2. use cached single-character ascii strings
+    // 3. if all bytes are ascii printable chars, don't do charset conversion
+    // 4. if available, use jvm no-copy private String constructor
```

Comment:
Neat! But have you compared the performance of this builder to the standard java (not scala) StringBuilder? @jcdavis found profound differences between them, in that the JVM has a ton of clever optimizations for the former.